dexlib2/src/main/java/org/jf/dexlib2/builder/LocatedItems.java

```diff
@@ -0,0 +1,89 @@
+package org.jf.dexlib2.builder;
+
+import com.google.common.collect.ImmutableList;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.*;
+
+public abstract class LocatedItems<T extends ItemWithLocation> {
+    @Nullable
+    private List<T> items = null;
+
+    @Nonnull
+    private List<T> getMutableItems() {
+        if (items == null) {
+            items = new ArrayList<>(1);
+        }
+        return items;
+    }
+
+    @Nonnull
+    private List<T> getImmutableItems() {
+        if (items == null) {
+            return ImmutableList.of();
+        }
+        return items;
+    }
+
+
+
+    public Set<T> getModifiableItems(MethodLocation newItemsLocation) {
+        return new AbstractSet<T>() {
+            @Nonnull
+            @Override public Iterator<T> iterator() {
+                final Iterator<T> it = getImmutableItems().iterator();
+
+                return new Iterator<T>() {
+                    private @Nullable
+                    T currentItem = null;
+
+                    @Override public boolean hasNext() {
+                        return it.hasNext();
+                    }
+
+                    @Override public T next() {
+                        currentItem = it.next();
+                        return currentItem;
+                    }
+
+                    @Override public void remove() {
+                        if (currentItem != null) {
+                            currentItem.setLocation(null);
+                        }
+                        it.remove();
+                    }
+                };
+            }
+
+            @Override public int size() {
+                return getImmutableItems().size();
+            }
+
+            @Override public boolean add(@Nonnull T item) {
+                if (item.isPlaced()) {
+
+
+                    throw new IllegalArgumentException(addLocatedItemError());
```

Comment:
maybe getAddLocatedItemError()? "addLocatedItemError" implies that the method is adding an error to something.